ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs

```diff
@@ -0,0 +1,257 @@
+namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Daemon.Stages
+
+open System
+open System.Collections.Generic
+open FSharp.Compiler.Symbols
+open JetBrains.Application.Settings
+open JetBrains.DocumentModel
+open JetBrains.ProjectModel
+open JetBrains.ReSharper.Feature.Services.Daemon
+open JetBrains.ReSharper.Plugins.FSharp
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Features.Daemon.Highlightings
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Features.Daemon.Stages
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Impl
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Tree
+open JetBrains.ReSharper.Plugins.FSharp.Settings
+open JetBrains.ReSharper.Psi.Tree
+open JetBrains.TextControl.DocumentMarkup.Adornments
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Services.Util.TypeAnnotationsUtil
+open JetBrains.TextControl.DocumentMarkup.Adornments.IntraTextAdornments
+open JetBrains.Util.Logging
+
+type private NodesRequiringHints =
+    { TopLevelNodes: List<ITreeNode>; LocalNodes: List<ITreeNode> }
+
+type private FSharpTypeHintSettings =
+    { TopLevelMembers: PushToHintMode; LocalBindings: PushToHintMode } with
+
+    static member Create(settingsStore: IContextBoundSettingsStore) =
+        { TopLevelMembers = settingsStore.GetValue(fun (key: FSharpTypeHintOptions) -> key.ShowTypeHintsForTopLevelMembers)
+                                         .EnsureInlayHintsDefault(settingsStore)
+          LocalBindings = settingsStore.GetValue(fun (key: FSharpTypeHintOptions) -> key.ShowTypeHintsForLocalBindings)
+                                       .EnsureInlayHintsDefault(settingsStore) }
+
+    member x.IsDisabled =
+        x.TopLevelMembers = PushToHintMode.Never &&
+        x.LocalBindings = PushToHintMode.Never
+
+
+type private MembersVisitor(settings) =
+    inherit TreeNodeVisitor<NodesRequiringHints>()
+
+    let isTopLevelMember (node: ITreeNode) =
+        match node with
+        | :? ITopBinding
+        | :? IMemberDeclaration
+        | :? IConstructorDeclaration -> true
+        | _ -> false
+
+    override x.VisitNode(node, context) =
+        if settings.LocalBindings = PushToHintMode.Never && isTopLevelMember node then () else
+
+        for child in node.Children() do
```

Comment:
Does it iterate through all the children when the current mode should only show the hints for the top level nodes? Or do we need to create the hints anyway?

Reply:
If we want to show hints only for the top level, then the check above will not allow us to go inside the top binding. We create hints for children if the Always/PushToHint mode is enabled for local bindings.